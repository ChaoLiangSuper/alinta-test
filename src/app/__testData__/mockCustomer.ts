export default {
  key: '111111',
  firstName: 'first',
  lastName: 'last',
  dateOfBirth: '10/27/2018',
};
