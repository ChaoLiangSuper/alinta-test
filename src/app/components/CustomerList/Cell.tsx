import styled from 'styled-components';

const Cell = styled.div<{ portion?: number }>`
  flex: ${({ portion = 1 }) => portion};
`;
export default Cell;
