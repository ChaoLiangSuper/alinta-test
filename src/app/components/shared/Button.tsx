import styled from 'styled-components';

interface ButtonProps {
  primary?: boolean;
}

const StyledButton = styled.button<ButtonProps>`
  border: none;
  padding: ${({ theme }) => theme.spacing()};
  border-radius: ${({ theme }) => theme.borderRadius()};
  background: ${({ theme, primary }) =>
    primary ? theme.primaryColor : theme.secondaryColor};
  color: ${({ theme, primary }) =>
    primary ? theme.primaryText : theme.secondaryText};

  &:active {
    background-color: ${({ theme, primary }) =>
      primary ? theme.primaryColorDark : theme.secondaryColorDark};
  }

  & + & {
    margin-left: ${({ theme }) => theme.spacing()};
  }
`;

export default StyledButton;
