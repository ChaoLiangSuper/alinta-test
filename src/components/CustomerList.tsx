import React from 'react';

const CustomerList: React.FC = () => {
  return <div>List</div>;
};

export default CustomerList;
