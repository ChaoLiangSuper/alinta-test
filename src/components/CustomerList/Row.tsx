import styled from 'styled-components';

const Row = styled.div`
  display: flex;
  width: 100%;
  border-bottom: 1px solid lightgray;
  padding: ${({ theme }) => theme.spacing(2)};

  &:hover {
    background: ${({ theme }) => theme.background};
  }
`;

export default Row;
