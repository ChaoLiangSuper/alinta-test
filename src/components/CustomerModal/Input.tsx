import React from 'react';
import styled from 'styled-components';

interface StyledInputProps {
  invalid?: boolean;
}

interface InputProps extends StyledInputProps {
  name: string;
  label: string;
  date?: boolean;
  value: string;
  onChange: (value: string) => void;
}

const StyledLabel = styled.label`
  display: flex;
  width: 100%;
`;

const StyledText = styled.span`
  flex-grow: 2;
  flex-basis: 0;
`;

const StyledInput = styled.input<StyledInputProps>`
  flex-grow: 3;
  flex-basis: 0;
  width: 100%;
  border: ${({ invalid }) => (invalid ? '1px solid red' : '1px solid grey')};
  border-radius: ${({ theme }) => theme.borderRadius()};
`;

const Input: React.FC<InputProps> = ({
  name,
  label,
  date,
  value,
  onChange,
  invalid,
}) => {
  return (
    <StyledLabel>
      <StyledText>{label}</StyledText>
      <StyledInput
        autoComplete='off'
        name={name}
        type={date ? 'date' : 'text'}
        value={value}
        invalid={invalid}
        onChange={({ target }) => onChange(target.value)}
      />
    </StyledLabel>
  );
};

export default Input;
