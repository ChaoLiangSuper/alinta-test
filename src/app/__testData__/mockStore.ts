import mockCustomer from './mockCustomer';
import { createStore } from 'redux';

export const storeData = {
  isLoading: false,
  customers: { [mockCustomer.key]: mockCustomer },
};

export default createStore(() => storeData);
