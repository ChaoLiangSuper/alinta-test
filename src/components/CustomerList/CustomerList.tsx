import _ from 'lodash';
import React from 'react';
import { connect } from 'react-redux';
import { Store } from '../../../types';
import Button from '../shared/Button';
import Row from './Row';
import Cell from './Cell';

interface CustomerListProps {
  customers: Store['customers'];
}

const CustomerList: React.FC<CustomerListProps> = ({ customers }) => {
  return (
    <>
      <Row>
        <Cell>First Name</Cell>
        <Cell>Last Name</Cell>
        <Cell>Date of Birth</Cell>
        <Cell>Actions</Cell>
      </Row>
      {_.isEmpty(customers) ? (
        <Row>No Customer Founded</Row>
      ) : (
        _.map(customers, ({ firstName, lastName, dateOfBirth, key }) => (
          <Row key={key}>
            <Cell>{firstName}</Cell>
            <Cell>{lastName}</Cell>
            <Cell>{dateOfBirth}</Cell>
            <Cell>
              <Button>Delete</Button>
              <Button primary>Edit</Button>
            </Cell>
          </Row>
        ))
      )}
    </>
  );
};

export default connect(({ customers }: Store) => ({ customers }))(CustomerList);
