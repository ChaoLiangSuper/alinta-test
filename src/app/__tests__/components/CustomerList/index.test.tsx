import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import Button from '../../../components/shared/Button';
import { CustomerList as Component } from '../../../components/CustomerList/CustomerList';
import { storeData } from '../../../__testData__/mockStore';
import Row from '../../../components/CustomerList/Row';
import Input from '../../../components/shared/Input';
import Cell from '../../../components/CustomerList/Cell';

describe('<CustomerList />', () => {
  let wrapper: ShallowWrapper;
  const mockFunc = jest.fn();

  beforeEach(() => {
    wrapper = shallow(
      <Component
        customers={storeData.customers}
        selectCustomer={mockFunc}
        deleteCustomer={mockFunc}
      />
    );
  });

  it('should render without error', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should show empty string', () => {
    wrapper.setProps({
      customers: {},
    });
    expect(
      wrapper
        .find(Row)
        .last()
        .text()
    ).toBe('No Customer Founded');
  });

  it('should be able to select customer', () => {
    wrapper.setProps({
      customers: {
        '123': {
          key: '123',
          firstName: 'first',
          lastName: 'last',
          dateOfBirth: '10/27/2018',
        },
      },
    });
    wrapper
      .find(Button)
      .last()
      .simulate('click');
    expect(mockFunc).toHaveBeenLastCalledWith('123');
  });

  it('should be able to delete customer', () => {
    wrapper.setProps({
      customers: {
        '124': {
          key: '124',
          firstName: 'first',
          lastName: 'last',
          dateOfBirth: '10/27/2018',
        },
      },
    });
    wrapper
      .find(Button)
      .first()
      .simulate('click');
    expect(mockFunc).toHaveBeenLastCalledWith('124');
  });

  it('should be able to search based on first name', done => {
    wrapper.setProps({
      customers: {
        '123': {
          key: '123',
          firstName: 'first1',
          lastName: 'last1',
          dateOfBirth: '10/27/2018',
        },
        '124': {
          key: '124',
          firstName: 'first2',
          lastName: 'last2',
          dateOfBirth: '10/27/2018',
        },
      },
    });
    wrapper.find(Input).prop('onChange')('first1');
    setTimeout(() => {
      expect(wrapper.find(Row).length).toBe(3);
      const lastRow = wrapper.find(Row).last();
      expect(
        lastRow
          .find(Cell)
          .first()
          .text()
      ).toBe('first1');
      done();
    }, 500);
  });

  it('should be able to search based on last name', done => {
    wrapper.setProps({
      customers: {
        '123': {
          key: '123',
          firstName: 'first1',
          lastName: 'last1',
          dateOfBirth: '10/27/2018',
        },
        '124': {
          key: '124',
          firstName: 'first2',
          lastName: 'last2',
          dateOfBirth: '10/27/2018',
        },
      },
    });
    wrapper.find(Input).prop('onChange')('last2');
    setTimeout(() => {
      expect(wrapper.find(Row).length).toBe(3);
      const lastRow = wrapper.find(Row).last();
      expect(
        lastRow
          .find(Cell)
          .first()
          .text()
      ).toBe('first2');
      done();
    }, 500);
  });
});
