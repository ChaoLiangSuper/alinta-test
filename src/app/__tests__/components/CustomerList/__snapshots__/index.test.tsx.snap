// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CustomerList /> should render without error 1`] = `
<Fragment>
  <styled.div>
    <styled.div>
      <Input
        label="Search"
        onChange={[Function]}
      />
    </styled.div>
    <styled.div
      portion={3}
    />
  </styled.div>
  <styled.div>
    <styled.div>
      First Name
    </styled.div>
    <styled.div>
      Last Name
    </styled.div>
    <styled.div>
      Date of Birth
    </styled.div>
    <styled.div>
      Actions
    </styled.div>
  </styled.div>
  <styled.div
    key="111111"
  >
    <styled.div>
      first
    </styled.div>
    <styled.div>
      last
    </styled.div>
    <styled.div>
      10/27/2018
    </styled.div>
    <styled.div>
      <styled.button
        onClick={[Function]}
      >
        Delete
      </styled.button>
      <styled.button
        onClick={[Function]}
        primary={true}
      >
        Edit
      </styled.button>
    </styled.div>
  </styled.div>
</Fragment>
`;
